sklearn/_fast_gradient_boosting/binning.py

```diff
@@ -0,0 +1,140 @@
+"""
+This module contains the BinMapper class.
+
+BinMapper is used for mapping a real-valued dataset into integer-valued bins.
+Bin thresholds are computed with the quantiles so that each bin contains
+approximately the same number of samples.
+"""
+# Author: Nicolas Hug
+
+import numpy as np
+
+from ..utils import check_random_state, check_array
+from ..base import BaseEstimator, TransformerMixin
+from ..utils.validation import check_is_fitted
+from ._binning import _map_to_bins
+from .types import X_DTYPE, X_BINNED_DTYPE
+
+
+def _find_binning_thresholds(data, max_bins=256, subsample=int(2e5),
+                             random_state=None):
+    """Extract feature-wise quantiles from numerical data.
+
+    Return
+    ------
+    binning_thresholds: tuple of arrays
+        For each feature, stores the increasing numeric values that can
+        be used to separate the bins. len(binning_thresholds) == n_features.
+    """
+    if not (2 <= max_bins <= 256):
+        raise ValueError('max_bins={} should be no smaller than 2 '
+                         'and no larger than 256.'.format(max_bins))
+    rng = check_random_state(random_state)
+    if subsample is not None and data.shape[0] > subsample:
+        subset = rng.choice(np.arange(data.shape[0]), subsample)
+        data = data[subset]
+
+    percentiles = np.linspace(0, 100, num=max_bins + 1)
+    end = percentiles.shape[0]  # no negative indexing!
+    percentiles = percentiles[1:end - 1]
+    binning_thresholds = []
+    for f_idx in range(data.shape[1]):
+        col_data = np.ascontiguousarray(data[:, f_idx], dtype=X_DTYPE)
+        distinct_values = np.unique(col_data)
+        if len(distinct_values) <= max_bins:
+            end = distinct_values.shape[0]  # no negative indexing!
+            midpoints = (distinct_values[:end - 1] + distinct_values[1:])
+            midpoints *= .5
+        else:
+            # We sort again the data in this case. We could compute
+            # approximate midpoint percentiles using the output of
+            # np.unique(col_data, return_counts) instead but this is more
+            # work and the performance benefit will be limited because we
+            # work on a fixed-size subsample of the full data.
+            midpoints = np.percentile(col_data, percentiles,
+                                      interpolation='midpoint').astype(X_DTYPE)
+        binning_thresholds.append(midpoints)
```

Comment:
this is a very expensive operation, and it matters if the data is of a high dimensionality. Pre-allocate it maybe?

Reply:
Are you sure? `append` should just store a ref to `midpoints`, not copy the whole array, so it should be cheap?

One small problem with pre-allocation is that the `midpoints` arrays have different dimensions (because of the number of unique values in each feature).

Reply:
This is not about what goes in the list, it's about `realloc`ing the list, which is pretty expensive. Even if you start with a list of `None`s and set the values to what you're creating in the `for` loop, it'd be faster. I've achieved speedups of the order 1000x at times with this trick. But you'd see that here only if there are many features and the list becomes large-ish.

Reply:
I'm a bit puzzled, `append` is supposed to be an amortized `O(1)` operation. I can't reproduce your speedup:

```python

n = 1_000_000

def f():
    l = []
    for i in range(n):
        l.append(i)
        
def g():
    l = [None] * n
    for i in range(n):
        l[i] = i


%timeit f()
92.4 ms ± 1.78 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit g()
60.5 ms ± 622 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Are you sure that you weren't in a scenario where the objects in the list were copied each time the list was expanded?

Reply:
You should not get more than x2: https://reterwebber.wordpress.com/2013/07/03/python-list-performance/